### PR TITLE
Task/um 696 allocator accessibility

### PR DESCRIPTION
**Pull request by kab163**
Notes:
1) Change required for HIP within blt stuff, but I couldn't include that modification in my branch - see comment in Umpire.cpp 
2) Table comment for is_accessible
3) Use of "UNIQUE" in example test - see comment in allocAccess.cpp file
4) Couldn't test HIP stuff because machine is still down, but it does compile at least

### CODE REVIEW

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
LGTM

File: src/umpire/Umpire.cpp [hunk 1]
```diff
@@ -98,6 +113,166 @@ bool pointer_contains(void* left_ptr, void* right_ptr)
...
+  switch(p) {
+    case (cPlatform::host):
+    {
+      if(findResource(a) == myResource::UNKNOWN
+       || findResource(a) == myResource::DEVICE_CONST)
+        return false;
+
+#if defined(UMPIRE_ENABLE_CUDA)
+      //check DEVICE case for (Umpire) CUDA
+      return is_host_pageable_from_cuda();
```
Comment (davidbeckingsale) [inline src/umpire/Umpire.cpp:182]: Can you merge these into one `is_host_pageable` function?

File: src/umpire/Umpire.hpp [hunk 2]
```diff
@@ -122,6 +122,21 @@ bool pointer_overlaps(void* left, void* right);
  */
 bool pointer_contains(void* left, void* right);
 
+/*!
+ * \brief Check whether or not an Allocator is accessible from a given platform.
+ *
+ *\param camp::Platform p  
+ *\param umpire::Allocator a 
+ */
+bool isAccessible(Platform p, Allocator a);
```
Comment (davidbeckingsale) [inline src/umpire/Umpire.hpp:131]:
```suggestion
bool is_accessible(Platform p, Allocator a);
```

Use snake_case for free functions.

**Comment by kab163** [comment]
LGTM

File: src/umpire/Umpire.hpp [hunk 3]
```diff
@@ -122,6 +122,21 @@ bool pointer_overlaps(void* left, void* right);
...
+
+/*!
+ * \brief Helper function to find the resource type of a given allocator
+ *
+ *\param umpire::allocator a  
+ */
+MemoryResourceTraits::resource_type findResource(Allocator a);
```
Comment (davidbeckingsale) [inline src/umpire/Umpire.hpp:138]: If you want to keep this public, can you rename to `get_resource`. I think that "find" makes it sound like it has to search for something, but we already know the resource.

Alternatively, you could move this into an anonymous namespace in Umpire.cpp so it won't be exposed to users.

**Review by davidbeckingsale — CHANGES_REQUESTED** [review]

File: src/umpire/Umpire.hpp (hunk 2 above)
Comment (kab163) [reply src/umpire/Umpire.hpp:131]: This is a good catch! I didn't see that. (And I think odd that it didn't give me a warning about it)

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
LGTM

File: src/umpire/Umpire.cpp [hunk 4]
```diff
@@ -13,8 +13,20 @@
...
 #include "umpire/ResourceManager.hpp"
+#include "umpire/util/MemoryResourceTraits.hpp"
+#include "camp/resource/platform.hpp"
 #include "umpire/config.hpp"
 
+#if defined(UMPIRE_ENABLE_CUDA)
+  #include <cuda_runtime_api.h>
+#endif
+
+//make sure "/opt/rocm-3.6.0/hsa/include/" is
```
Comment (kab163) [inline src/umpire/Umpire.cpp:24]: Not sure how or if i can edit the blt stuff - is this the best I can do for this?

File: tests/integration/allocator_accessibility.cpp [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+#include "gtest/gtest.h"
+#include "gtest/gtest-death-test.h"
+#include "umpire/Allocator.hpp"
+#include "umpire/ResourceManager.hpp"
+#include "umpire/Umpire.hpp"
+#include "umpire/config.hpp"
+#include "umpire/util/MemoryResourceTraits.hpp"
+
+//Need to enable gtest death tests with 
+// -DENABLE_GTEST_DEATH_TESTS=On
```
Comment (kab163) [inline tests/integration/allocator_accessibility.cpp:16]: This is another thing I'm unsure about. Whenever this test is added to the CMakeLists.txt, this flag must be turned on. Maybe we can add this flag to the Developer Defaults? Or something else to make it automatic (I don't think the tests will run currently)

File: tests/integration/allocator_accessibility.cpp [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    else {
+      double* data = static_cast<double*>(m_allocator->allocate(m_size * sizeof(double)));
+      ASSERT_DEATH(test(data, m_size, dev), "");
+    }
+  }
+}
+#endif
+
+/////////////////////////////
+//Sycl test not yet available
```
Comment (kab163) [inline tests/integration/allocator_accessibility.cpp:163]: This could be left as is, or I could mimic the OpenMP Target test just so that something is here. For now, there's just this placeholder.

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
LGTM for the win!!

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
Plz work

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
LGTM for the win!!

**Comment by kab163** [comment]
Documenting our CUDA and MPI problem:
Tests with CUDA work if ran with  lrun -n 1 --smpiargs="-disable_gpu_hooks" ./bin/allocator_accessibility_tests
More info here:
https://docs.olcf.ornl.gov/systems/summit_user_guide.html#cuda-hook-error-when-program-uses-cuda-without-first-calling-mpi-init

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
LGTM

File: src/umpire/ResourceManager.hpp [hunk 7]
```diff
@@ -84,6 +84,17 @@ class ResourceManager {
    */
   Allocator getDefaultAllocator();
 
+  /*!
+   * \brief Get the existing Resources.
```
Comment (mcfadden8) [inline src/umpire/ResourceManager.hpp:88]:
```suggestion
   * \brief Get the names for existing Resources.
```

File: src/umpire/Umpire.cpp (hunk 4 above)
Comment (mcfadden8) [reply src/umpire/Umpire.cpp:24]: Is this a change that is required in blt?  If so, has an issue been written up there?  I'm not sure these comments belong in the source code.

File: src/umpire/Umpire.cpp [hunk 8]
```diff
@@ -98,6 +113,141 @@ bool pointer_contains(void* left_ptr, void* right_ptr)
...
 
+bool is_host_pageable()
+{
+#if defined(UMPIRE_ENABLE_HIP)
+  hipDeviceProp_t props;
+  int hdev = 0;
+  hipGetDevice(&hdev);
+
+  //Check whether HIP can map host memory.
+  hipGetDeviceProperties(&props, hdev);
```
Comment (mcfadden8) [inline src/umpire/Umpire.cpp:124]: Can these device-implementation-specific calls be placed in to static methods of the specific resource instead of being placed here?

File: examples/allocator_access.cpp [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+#include "umpire/Allocator.hpp"
+#include "umpire/ResourceManager.hpp"
+#include "umpire/Umpire.hpp"
+
+using cPlatform = camp::resources::Platform;
+
+bool testAccess(umpire::Allocator a)
+{
+  if(is_accessible(cPlatform::host, a)) {
```
Comment (mcfadden8) [inline examples/allocator_access.cpp:18]: Shouldn't this call have a namespace qualifier?  I am curious why this compiles.  Is there a `using umpire` somewhere?  if so, that doesn't seem right to me.

```suggestion
  if(umpire::is_accessible(cPlatform::host, a)) {
```

File: tests/integration/allocator_accessibility.cpp [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    delete m_allocator;
+  }
+  
+  umpire::Allocator* m_allocator;
+  size_t m_size = 42;
+};
+
+TEST_P(AllocatorAccessibilityTest, AccessibilityFromPlatform)
+{
+  if(is_accessible(cPlatform::host, *m_allocator)) {
```
Comment (mcfadden8) [inline tests/integration/allocator_accessibility.cpp:119]: Similar question to another comment made in a different file.  Shouldn't this have a namespace qualified?  I'm confused why it compiles.
```suggestion
  if(umpire::is_accessible(cPlatform::host, *m_allocator)) {
```

**Review by mcfadden8 — CHANGES_REQUESTED** [review]

File: src/umpire/ResourceManager.hpp (hunk 7 above)
Comment (kab163) [reply src/umpire/ResourceManager.hpp:88]: Oooo good catch!

File: src/umpire/Umpire.cpp (hunk 4 above)
Comment (kab163) [reply src/umpire/Umpire.cpp:24]: You're right. I wasn't sure how to handle this, but if I didn't add that path, the "hsa/hsa.h" file couldn't be found (needed for hip stuff). I think it's a blt change requirement, but I'm not sure how to make edits to the blt submodule stuff. I agree that the comment doesn't belong there, it was kind of a placeholder for figuring it out later lol

@davidbeckingsale any thoughts?

Comment (davidbeckingsale) [reply src/umpire/Umpire.cpp:24]: You can now set the CMake variable HIP_RUNTIME_INCLUDE_DIRS - take a look here for an example: https://github.com/LLNL/Umpire/blob/develop/Dockerfile#L69

Comment (kab163) [reply src/umpire/Umpire.cpp:24]: Great, I'll remove this comment then. Thanks!

File: examples/allocator_access.cpp (hunk 9 above)
Comment (kab163) [reply examples/allocator_access.cpp:18]: The is_accessible function is housed in the Umpire.hpp file, which is included. I assume that's why it doesn't complain of an error. I could try to do some kind of statement like `using access = <path to is_accessible function>` similar to how I did the `cPlatform` statement instead (I'd have to test it, haven't tried as of yet). Is one better than the other?

File: tests/integration/allocator_accessibility.cpp (hunk 10 above)
Comment (kab163) [reply tests/integration/allocator_accessibility.cpp:119]: The umpire.hpp file is included, I assume that's why this works. The is_accessible function is not a member of the umpire namespace, so if I did something like that, it'd have to make it something like in line 14 or 15. Am I explaining this right? I feel like I'm making this confusing, but do you see what I'm talking about?

File: src/umpire/Umpire.cpp (hunk 8 above)
Comment (davidbeckingsale) [reply src/umpire/Umpire.cpp:122]: They cannot be static since some info (e.g. device number) would be a property of a specific resource instance.

File: tests/integration/allocator_accessibility.cpp (hunk 10 above)
Comment (mcfadden8) [reply tests/integration/allocator_accessibility.cpp:119]: Oh, yes.  I get it.  I think that these functions need to be part of the umpire namespace.

File: src/umpire/Umpire.cpp (hunk 8 above)
Comment (kab163) [reply src/umpire/Umpire.cpp:122]: Good idea, stay tuned! :)

File: tests/integration/allocator_accessibility.cpp (hunk 10 above)
Comment (davidbeckingsale) [reply tests/integration/allocator_accessibility.cpp:119]: The function is in the `umpire` namespace. I think there might a `using namespace umpire` line somewhere that makes this work.

Comment (davidbeckingsale) [reply tests/integration/allocator_accessibility.cpp:119]: Adding a `using` declaration like on line 14 or 15 is not necessary if you explicitly write out the fully qualified name.

Comment (mcfadden8) [reply tests/integration/allocator_accessibility.cpp:119]: If there is a `using umpire` statement somewhere, I would like to get it removed (especially if it is in a header).  using namespaces should be discouraged in libraries.

Comment (kab163) [reply tests/integration/allocator_accessibility.cpp:119]: Oh, well Umpire.hpp has the umpire namespace, maybe when I include that file, it kinda gets added in I guess

Comment (mcfadden8) [reply tests/integration/allocator_accessibility.cpp:119]: uh oh.  IMHO, that is a problem.

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
TODO: make default to return false instead of returning true in isAccessible() functions; double check cases; update Accessibility Table documentation so there's a device-hip, device-cuda, etc. version.

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
LGTM

File: scripts/uberenv/packages/umpire/package.py [hunk 11]
```diff
@@ -319,6 +319,8 @@ def hostconfig(self, spec, prefix, py_site_pkgs_dir=None):
                                         hip_root))
             cfg.write(cmake_cache_entry("HIP_CLANG_PATH",
                                         rocm_root + '/llvm/bin'))
+            cfg.write(cmake_cache_entry("HIP_HIPCC_FLAGS",
+                                        '--amdgpu-target=gfx906'))
```
Comment (mcfadden8) [inline scripts/uberenv/packages/umpire/package.py:323]: Was this change required for the allocator accessibility check or was it needed for something else?

File: src/umpire/Umpire.hpp [hunk 12]
```diff
@@ -122,6 +122,19 @@ bool pointer_overlaps(void* left, void* right);
...
+ *
+ * This function describes which allocators should be accessible
+ * from which CAMP platforms. Information on platform/allocator
+ * accessibility can be found at
+ * <https://umpire.readthedocs.io/en/develop/features/allocator_accessibility.html>
+ *
+ *\param camp::Platform p  
+ *\param umpire::Allocator a 
+ */
+bool is_accessible(Platform p, Allocator a);
```
Comment (mcfadden8) [inline src/umpire/Umpire.hpp:136]: Shouldn't this be:
```suggestion
bool is_accessible(camp::Platform p, Allocator a);
```

File: scripts/uberenv/packages/umpire/package.py (hunk 11 above)
Comment (kab163) [reply scripts/uberenv/packages/umpire/package.py:323]: It was needed so that when we are from the perspective of the host platform, for example, we can ask if the right device resource is accessible (I think like the main device, not DEVICE::1 or something). When this flag wasn't included, I got some odd HIP errors about device usage. This just makes it so that the flag is always included.

File: src/umpire/Umpire.cpp [hunk 13]
```diff
@@ -98,6 +99,20 @@ bool pointer_contains(void* left_ptr, void* right_ptr)
   }
 }
 
+bool is_accessible(Platform p, Allocator a)
```
Comment (mcfadden8) [inline src/umpire/Umpire.cpp:102]: I'm not sure how the `camp` namespace is being made available.  But since this is a library function, shouldn't the camp namespace be explicitly stated?
```suggestion
bool is_accessible(camp::Platform p, Allocator a) 
```

File: src/umpire/Umpire.cpp [hunk 14]
```diff
@@ -12,6 +12,7 @@
 #include <iterator>
 #include <sstream>
 
+#include "umpire/resource/MemoryResource.hpp"
 #include "umpire/ResourceManager.hpp"
 #include "umpire/config.hpp"
```
Comment (mcfadden8) [inline src/umpire/Umpire.cpp:17]: Shouldn't the implementation file also include the camp.hpp file?

File: src/umpire/Umpire.hpp (hunk 12 above)
Comment (kab163) [reply src/umpire/Umpire.hpp:136]: I include the Platform.hpp file, so "Platform p" is really doing "camp::resources::Platform p" - I don't think it's needed, but I may be wrong.

Comment (mcfadden8) [reply src/umpire/Umpire.hpp:136]: Oh.  I see.  The comment at line 133 says "camp::Platform".   Platform, umpire::Platform, and camp::Platform are going to be confusing.  I think Platform.hpp existed before camp came around.  My personal opinion is we should somehow deprecate Platform.hpp and instead use camp::Platform.

File: src/umpire/alloc/MallocAllocator.hpp [hunk 15]
```diff
@@ -50,6 +58,47 @@ struct MallocAllocator {
     UMPIRE_LOG(Debug, "(ptr=" << ptr << ")");
     ::free(ptr);
   }
+
+  bool isHostPageable()
+  {
+#if defined(UMPIRE_ENABLE_HIP)
+    hipDeviceProp_t props;
+    int hdev = 0;
+    hipGetDevice(&hdev);
```
Comment (mcfadden8) [inline src/umpire/alloc/MallocAllocator.hpp:67]: Similar to what was done with device-specific calls in ResourceManager, can these calls also be moved down into the resource?

Comment (mcfadden8) [reply src/umpire/alloc/MallocAllocator.hpp:67]: Or are we saying that things in umpire/alloc are the same as things in umpire/resources?  (@davidbeckingsale , I am not sure?)

Comment (davidbeckingsale) [reply src/umpire/alloc/MallocAllocator.hpp:67]: They have to stay here.

File: src/umpire/Umpire.cpp (hunk 13 above)
Comment (davidbeckingsale) [reply src/umpire/Umpire.cpp:102]: We bring the Platform enum into the umpire namespace too.

**Review by mcfadden8 — CHANGES_REQUESTED** [review]
I think this is looking great!  I especially like the way the implementation looks in ResourceManager.cpp.  I just had a few questions.

File: scripts/uberenv/packages/umpire/package.py (hunk 11 above)
Comment (davidbeckingsale) [reply scripts/uberenv/packages/umpire/package.py:323]: It's required because the accessibility tests generate device code. In the past we were only calling the HIP API, so this flag wasn't needed.

**Review by mcfadden8 — APPROVED** [review]

File: src/umpire/alloc/MallocAllocator.hpp (hunk 15 above)
Comment (kab163) [reply src/umpire/alloc/MallocAllocator.hpp:67]: I think this is the best it can get. So this file is looking from the host resource memory perspective and asking if the cuda/hip platform is accessible. So this technically is down in the resource, but because i have to ask from the host resource if it can access the device platforms, I have to use the device-specific calls. If you look at the HipDeviceMemoryResource.cpp file, you'll see an example of how I ask, from the device memory resource perspective, if the host platform is accessible (so the reverse of this file, in regard to hip). 

This is some very detailed stuff, so we can have a webex or something to discuss it if you'd like.

Comment (kab163) [reply src/umpire/alloc/MallocAllocator.hpp:67]: @davidbeckingsale, any thoughts?

File: src/umpire/Umpire.hpp (hunk 12 above)
Comment (kab163) [reply src/umpire/Umpire.hpp:136]: Wait, maybe I don't lol let me look..

File: src/umpire/alloc/MallocAllocator.hpp (hunk 15 above)
Comment (davidbeckingsale) [reply src/umpire/alloc/MallocAllocator.hpp:67]: I think we should discuss this next Tuesday, but I'm happy with how this works right now. This isn't resource-specific since a camp::Platform doesn't define things like the device number, so we just have to do the best way can to determine whether or not the host memory is accessible from that platform.

File: src/umpire/Umpire.hpp (hunk 12 above)
Comment (davidbeckingsale) [reply src/umpire/Umpire.hpp:136]: We do use camp::Platform, but umpire::Platform is an alias.

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
TODO: find out more about why hip is_accessible() function fix was necessary; change the getResourceNames() func. in ResourceManager()

**Comment by kab163** [comment]
LGTM

File: src/umpire/ResourceManager.hpp [hunk 16]
```diff
@@ -84,6 +84,17 @@ class ResourceManager {
...
   Allocator getDefaultAllocator();
 
+  /*!
+   * \brief Get the names for existing Resources.
+   *
+   * The Resource Manager dynamically populates available memory resource types with
+   * m_resource_names based on what's available. This is used for determining 
+   * allocator accessibility.
+   *
+   * \return The private member, m_resource_names.
```
Comment (davidbeckingsale) [inline src/umpire/ResourceManager.hpp:94]: This comment is out of date now.

File: src/umpire/resource/HipDeviceMemoryResource.cpp [hunk 17]
```diff
@@ -65,6 +65,30 @@ std::size_t HipDeviceMemoryResource::getHighWatermark() const noexcept
...
+  else
+    return false;
+}
+
+bool HipDeviceMemoryResource::isAccessibleFrom(Platform p) noexcept
+{
+  if(p == Platform::hip)
+    return true;
+  else if(p == Platform::host)
+    return isHostPageable();
```
Comment (davidbeckingsale) [inline src/umpire/resource/HipDeviceMemoryResource.cpp:87]: I don't think you need to check this. As far as I know, HIP device memory will always be accessible from the host.

**Review by davidbeckingsale — CHANGES_REQUESTED** [review]
Couple of quick comments.

**Comment by kab163** [comment]
LGTM

File: src/umpire/resource/HipDeviceMemoryResource.cpp (hunk 17 above)
Comment (kab163) [reply src/umpire/resource/HipDeviceMemoryResource.cpp]: I updated this check so that if its the host or hip it will return true, otherwise return false

**Comment by kab163** [comment]
LGTM

File: src/umpire/alloc/MallocAllocator.hpp [hunk 18]
```diff
@@ -50,6 +58,35 @@ struct MallocAllocator {
...
+    //Device supports coherently accessing pageable memory
+    //without calling cudaHostRegister on it
+    cudaDeviceGetAttribute(&pageableMem,
+              cudaDevAttrPageableMemoryAccess, cdev);
+    if(pageableMem)
+      return true;
+    else
+      return false;
+#endif
+    return false; //shouldn't reach this
```
Comment (mcfadden8) [inline src/umpire/alloc/MallocAllocator.hpp:78]: You could actually drop the "else return false" from lines 75 and 76 above and the code would just fall through to this line and return false if not pageable or UMPIRE_ENABLE_CUDA is not defined.  Then you could also remove the comment and not have to worry about knowing that the only time this function is called is if the platform is Platform::cuda.

File: tests/integration/allocator_accessibility.cpp [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+#include "gtest/gtest.h"
+#include "umpire/Allocator.hpp"
+#include "umpire/ResourceManager.hpp"
+#include "umpire/Umpire.hpp"
+#include "umpire/config.hpp"
+#include "umpire/util/MemoryResourceTraits.hpp"
+#include "umpire/strategy/QuickPool.hpp"
+
+//using cPlatform = camp::resources::Platform;
+//using myResource = umpire::MemoryResourceTraits::resource_type;
```
Comment (mcfadden8) [inline tests/integration/allocator_accessibility.cpp:16]: Should these two comments be removed?

**Review by mcfadden8 — COMMENTED** [review]
Looking good.  Some minor changes requested.  Also, the gitlab CI builds need to be fixed before merge.

File: tests/integration/allocator_accessibility.cpp (hunk 19 above)
Comment (kab163) [reply tests/integration/allocator_accessibility.cpp:16]: done

File: src/umpire/alloc/MallocAllocator.hpp (hunk 18 above)
Comment (kab163) [reply src/umpire/alloc/MallocAllocator.hpp:78]: good idea, done!

File: examples/allocator_access.cpp [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+//
+// SPDX-License-Identifier: (MIT)
+//////////////////////////////////////////////////////////////////////////////
+#include <iostream>
+#include <string>
+#include "umpire/Allocator.hpp"
+#include "umpire/ResourceManager.hpp"
+#include "umpire/Umpire.hpp"
+
+bool testAccess(umpire::Allocator a)
```
[the remark this answers is not in the document]
Comment (kab163) [reply examples/allocator_access.cpp:13]: very true, done.

File: examples/allocator_access.cpp [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+
+  ///////////////////////////////////////////////////
+  //Test accessibility
+  ///////////////////////////////////////////////////
+  std::cout << "Testing the available allocators for "
+            << "accessibility from the CAMP host platform:" 
+            << std::endl << std::endl;
+  const int size = 100;
+  for(auto a : alloc) {
+    if(testAccess(a)) {
```
[the remark this answers is not in the document]
Comment (kab163) [reply examples/allocator_access.cpp:60]: done!

File: examples/allocator_access.cpp [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+      int* data = static_cast<int*>(a.allocate(size*sizeof(int)));
+      for(int i = 0; i < size; i++) {
+        data[i] = i * i;
+      }
+      std::cout << "Testing: " << data[size-1] << " should be equal to " 
+                << (size-1)*(size-1) << std::endl << std::endl; 
+    }
```
[the remark this answers is not in the document]
Comment (kab163) [reply examples/allocator_access.cpp:67]: good idea again, done!

**Comment by kab163** [comment]
LGTM

**Review by mcfadden8 — APPROVED** [review]
Looks good to me.  Thanks!

**Review by davidbeckingsale — APPROVED** [review]
